space-framework/pkg/msclientlib/client.go

```diff
@@ -30,44 +29,45 @@ import (
 	"k8s.io/client-go/tools/cache"
 	"k8s.io/client-go/tools/clientcmd"
 
-	mcclientset "github.com/kubestellar/kubestellar/pkg/spaceclient/clientset"
 	spacev1alpha1 "github.com/kubestellar/kubestellar/space-framework/pkg/apis/space/v1alpha1"
-	ksclientset "github.com/kubestellar/kubestellar/space-framework/pkg/client/clientset/versioned"
+	mgtclientset "github.com/kubestellar/kubestellar/space-framework/pkg/client/clientset/versioned"
 	ksinformers "github.com/kubestellar/kubestellar/space-framework/pkg/client/informers/externalversions"
 )
 
 const defaultProviderNs = "spaceprovider-default"
 
 type KubestellarSpaceInterface interface {
 	// Space returns clientset for given space.
-	Space(name string, namespace ...string) (client mcclientset.Interface)
+
+	//Space(name string, namespace ...string) (client mcclientset.Interface)
+
 	// ConfigForSpace returns rest config for given space.
 	ConfigForSpace(name string, namespace ...string) (*rest.Config, error)
 }
 
 type multiSpaceClient struct {
 	ctx              context.Context
 	configs          map[string]*rest.Config
-	clientsets       map[string]*mcclientset.Clientset
-	managerClientset *ksclientset.Clientset
+	managerClientset *mgtclientset.Clientset
 	lock             sync.Mutex
 }
 
-func (mcc *multiSpaceClient) Space(name string, namespace ...string) mcclientset.Interface {
-	key, ns := namespaceKey(name, namespace)
-	var err error
-	mcc.lock.Lock()
-	defer mcc.lock.Unlock()
-	clientset, ok := mcc.clientsets[key]
-	if !ok {
-		// Try to get Space from API server.
-		if _, clientset, err = mcc.getFromServer(name, ns); err != nil {
-			panic(fmt.Sprintf("invalid space name: %s. error: %v", name, err))
+/*
+	 func (mcc *multiSpaceClient) Space(name string, namespace ...string) mcclientset.Interface {
+		key, ns := namespaceKey(name, namespace)
+		var err error
+		mcc.lock.Lock()
+		defer mcc.lock.Unlock()
+		clientset, ok := mcc.clientsets[key]
+		if !ok {
+			// Try to get Space from API server.
+			if _, clientset, err = mcc.getFromServer(name, ns); err != nil {
+				panic(fmt.Sprintf("invalid space name: %s. error: %v", name, err))
+			}
 		}
+		return clientset
 	}
-	return clientset
-}
-
+*/
```

Comment:
Before this is really usable there will at least be a comment explaining why `ConfigForSpace` takes any number of namespace names.  Perhaps the design will change to something more definite.

Reply:
This was done to support 0 or 1 not any number. Is there another way to allow optional parameter ?

Reply:
1.  It is not enough to say that a parameter is optional.  The caller needs to understand the consequences of eliding the parameter, and needs to understand what values are reasonable to supply, and what are the consequences of doing that.  These are not obvious to me.  The name "ConfigForSpace" does not make it obvious why a namespace is relevant, let alone optional.
2. Other ways to have an optional parameter include (1) taking a pointer that might be nil, and (2) having a sentinel value that means omission (this is particularly natural when the zero value of the parameter type is never a valid value).